Add environment variables to define Android keystore information

Calinou opened the pull request:
These make it possible to specify keystore information without storing it into `export_presets.cfg`. When used, this makes `export_presets.cfg` safely checkable into version control. This should make setting up continuous integration significantly easier.

The variables are:

- `GODOT_ANDROID_KEYSTORE_PATH`: Absolute path to the keystore file
- `GODOT_ANDROID_KEYSTORE_USER`: Keystore username
- `GODOT_ANDROID_KEYSTORE_PASSWORD`: Keystore password

These will be used when exporting in both debug and release modes. If a variable isn't set, the old behavior will apply.

I'm marking this as a draft as it needs testing (cc @Myran) and better signalling in the editor GUI (if it's set before starting the graphical editor). Right now, if one of those variables is set, the editor Export dialog will still display the keystore properties as usual, even though it won't use their values. Any suggestions? I'd like to have the variables used in both headless and non-headless versions, as it's more consistent and can be useful to avoid storing the release keystore in the project files when working with version control.

This closes #2759. This partially addresses https://github.com/godotengine/godot-proposals/issues/1156. See also https://github.com/github/gitignore/pull/2827 and https://github.com/godotengine/godot-demo-projects/issues/329.

Calinou commented:
@Myran and others, you can use this Linux headless binary for testing: https://0x0.st/iiHW.zip (link will expire in late 2020)

Myran commented:
> @Myran and others, you can use this Linux headless binary for testing: https://0x0.st/iiHW.zip (link will expire in late 2020)
> 
> It was built on Ubuntu 16.04.

Unfortunately I use MacOS currently due to focus on both iOS and Android (A headless option for mac would be nice since you cant supress the window like on windows).
 
While this is good there are other parts of the export_preset.cfg that needs to be exposed in other cases. Maybe a more general solution for the whole export_preset is viable? 

Example: Currently to deploy new apk/ipa:s to testflight and play store, the version needs to be incremented each build. I use fastlane for iOS and there is support for incrementing the builds trough xcode however for apk:s I had to create this script:
```
extends SceneTree

func _init():
	var export_config = ConfigFile.new()
	var export_config_path = "res://export_presets.cfg"
	var config_error = export_config.load(export_config_path)
	print("Current version value:",export_config.get_value("preset.1.options","version/code"))
	var arguments = {}
	for argument in OS.get_cmdline_args():
		# Parse valid command-line arguments into a dictionary
		if argument.find("=") > -1:
			var key_value = argument.split("=")
			arguments[key_value[0].lstrip("--")] = key_value[1]
	var version = arguments["export_android_version"]
	print("Set version to:",version)
	export_config.set_value("preset.1.options","version/code",version)
	export_config.save(export_config_path)
	quit()
```
And run this before export.
```
godot.osx.opt.tools.64 --path project -s update_android_version.gd --export_android_version=$CI_PIPELINE_IID
```

Its currently very rough and I was planning to clean it up and generalize it to all settings when need arises (keystore being one of them) but if its possible to handle it internally it would be great.

Calinou commented:
> Unfortunately I use MacOS currently due to focus on both iOS and Android (A headless option for mac would be nice since you cant supress the window like on windows).

There is one, but you need to build Godot from source to use it. Just build Godot with `platform=server` as you'd do on Linux; it will give you a macOS headless binary.

Myran commented:
> > (A headless option for mac would be nice since you cant supress the window like on windows).
> 
> There is one, but you need to build Godot from source to use it. Just build Godot with `platform=server` as you'd do on Linux; it will give you a macOS headless binary.

I tried to build this a while ago but server wouldnt show as an option for platforms and it was implied somwhere that it was only for linux. But now it works without problems, thanks!

asheraryam commented:
Are there any blockers with this? I think what it accomplishes is good enough to solve the CI problem for pc/android.

Let me know if there's anything I can help with.

idoby commented:
@Calinou Hi, how can I help get this shipped? I can resolve the merge conflict and push to your repo if you want me to.

Calinou commented:
@idoby I just rebased it. Note that the `master` branch is currently undergoing refactoring, so getting this PR merged may take a while.

idoby commented:
I need it for my CI process too and I don't want to build Godot during CI.
I'd like at least some indication in the UI in case things stop working because I've set the env vars and I have to start figuring out why, but I agree this PR should get merged before that.
Thanks for your work, friend!

Calinou commented:
@idoby In the meantime, you can use `sed` commands to replace text in `export_presets.cfg` before exporting the project. It's not pretty, but it works :slightly_smiling_face: 

There are some examples of this in the wild, such as https://github.com/dbrova15/godot-ci-android-bakz/blob/master/.gitlab-ci.yml#L53.

idoby commented:
Thanks, but looks like I won't be able to get away with vanilla Godot anyway, since we need some modifications.
But anyway this is way off topic. Feel free to @ me if you need help with this PR or any other!

Calinou committed:
Add environment variables to define Android keystore information

These make it possible to specify keystore information without
storing it into `export_presets.cfg`. When this feature is used,
this makes `export_presets.cfg` safely checkable into version control.

The variables are:

- `GODOT_ANDROID_KEYSTORE_PATH`: Absolute path to the keystore file
- `GODOT_ANDROID_KEYSTORE_USER`: Keystore username
- `GODOT_ANDROID_KEYSTORE_PASSWORD`: Keystore password

These will be used when exporting in both debug and release modes.
If a variable isn't set, the old behavior will apply.

This closes #2759.

Calinou commented:
I remade the pull request from scratch for the new Android exporting code.

The UI now uses a different error message if the keystore path was specified using the environment variable.

m4gr3d commented on platform/android/export/export.cpp:
```diff
@@ -1993,21 +2002,38 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
+		String debug_keystore;
+		if (OS::get_singleton()->has_environment(ENV_KEYSTORE_PATH)) {
+			debug_keystore = OS::get_singleton()->get_environment(ENV_KEYSTORE_PATH);
+			if (!debug_keystore.is_empty() && !FileAccess::exists(debug_keystore)) {
```
Seems like we should fallback to the editor-wide keystore path, or at least print a warning if `ENV_KEYSTORE_PATH` is present but empty.

m4gr3d commented on platform/android/export/export.cpp:
```diff
@@ -2234,9 +2260,27 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
-		String release_keystore = p_preset->get("keystore/release");
-		String release_username = p_preset->get("keystore/release_user");
-		String release_password = p_preset->get("keystore/release_password");
+
+		String release_keystore;
+		if (OS::get_singleton()->has_environment(ENV_KEYSTORE_PATH)) {
+			release_keystore = OS::get_singleton()->get_environment(ENV_KEYSTORE_PATH);
+		} else {
+			release_keystore = p_preset->get("keystore/release");
+		}
+
+		String release_username;
+		if (OS::get_singleton()->has_environment(ENV_KEYSTORE_USER)) {
+			release_username = OS::get_singleton()->get_environment(ENV_KEYSTORE_USER);
+		} else {
+			release_username = p_preset->get("keystore/release_user");
+		}
+
+		String release_password;
+		if (OS::get_singleton()->has_environment(ENV_KEYSTORE_PASSWORD)) {
+			release_password = OS::get_singleton()->get_environment(ENV_KEYSTORE_PASSWORD);
+		} else {
+			release_password = p_preset->get("keystore/release_password");
+		}
```
We are repeating similar logics here, above and below. Can the retrieval logic be wrapped into a helper function instead. This would ensure that when we update on section, we don't introduce bugs by omitting to update the other sections.
The helper function could return a `struct` containing the `keystore_path`, `keystore_username` and `keystore_password`, as well as an error indicating if the returned information is valid or not.

nathanfranke commented:
I don't consider this a fix of https://github.com/godotengine/godot-proposals/issues/1156. From what I see, keystore passwords are still stored in plain text when configured through the editor and are by default committed to git repositories. I would argue this is a critical security flaw.

akien-mga commented:
No, this is fully superseded by https://github.com/godotengine/godot/pull/76165.
